src/binary.rs

```diff
@@ -0,0 +1,322 @@
+            return Ok(false);
+        }
+    }
+
+    Ok(true)
+}
+
+/// Download a file from its remote location in the given version, extract it and make it ready for
+// execution at the given location.
+#[tracing::instrument(level = "trace")]
+async fn download(app: Application, version: &str) -> Result<PathBuf> {
+    tracing::info!(version = version, "downloading {}", app.name());
+
+    let cache_dir = cache_dir().context("failed getting the cache directory")?;
+    let client = surf::client().with(Redirect::new(5));
+
+    let mut resp = client
+        .get(app.url(version)?)
+        .send()
+        .await
+        .map_err(|e| e.into_inner())
+        .context("error sending HTTP request")?;
+    ensure!(
+        resp.status().is_success(),
+        "error downloading binary file: {:?}\n{}",
+        resp.status(),
+        app.url(version)?
+    );
+
+    let temp_out = cache_dir.join(format!("{}-{}.tmp", app.name(), version));
+    let mut file = AsyncFile::create(&temp_out).await.context("failed creating temporary output file")?;
+
+    io::copy(resp.take_body().into_reader(), &mut file)
+        .await
+        .context("failed downloading the archive")?;
+
+    Ok(temp_out)
+}
+
+/// Install an application from a downloaded archive locating and copying it to the given target
+/// location.
+#[tracing::instrument(level = "trace")]
+fn install(app: Application, archive_file: &mut File, target: &Path) -> Result<()> {
+    tracing::info!("installing {}", app.name());
+
+    let mut archive = TarArchive::new(GzDecoder::new(archive_file));
+    let mut file = extract_file(&mut archive, target, Path::new(app.path()))?;
+
+    set_executable_flag(&mut file)?;
+
+    for path in app.extra_paths() {
+        // archive must be opened for each entry as tar files don't allow jumping forth and back.
+        let archive_file = archive.into_inner().into_inner();
+        archive_file.seek(SeekFrom::Start(0))?;
+
+        archive = TarArchive::new(GzDecoder::new(archive_file));
+        extract_file(&mut archive, target, Path::new(path))?;
+    }
+
+    Ok(())
+}
+
+/// Extract a single file from the given archive and put it into the target location.
+fn extract_file<R>(archive: &mut TarArchive<R>, target: &Path, file: &Path) -> Result<File>
+where
+    R: Read,
+{
+    let mut tar_file = find_tar_entry(archive, file)?.context("file not found in archive")?;
+    let out = target.join(file);
+
+    if let Some(parent) = out.parent() {
+        std::fs::create_dir_all(parent).context("failed creating output directory")?;
+    }
+
+    let mut out = File::create(target.join(file)).context("failed creating output file")?;
+    std::io::copy(&mut tar_file, &mut out).context("failed copying over final output file from archive")?;
+
+    Ok(out)
+}
+
+/// Locate the cache dir for trunk and make sure it exists.
+fn cache_dir() -> Result<PathBuf> {
+    let path = ProjectDirs::from("dev", "trunkrs", "trunk")
```

Comment:
`trunkrs` is really only a fallback due to not having access to the `trunk.rs` domain (it was already taken), and I probably should have made the site `trunk-rs.dev` lolz. Maybe we should just keep it as `("dev", "trunk")`. Thoughts?

Reply:
I see. This function [needs 3 arguments](https://docs.rs/directories-next/2.0.0/directories_next/struct.ProjectDirs.html#method.from) unfortunately (mostly due to MacOS) but as the docs say they can be empty as well.

How about `com.github`, `thedodd`, `trunk` instead?

Reply:
Hmm, I see. Thanks for the info. Yea, maybe `trunk-rs`. I have this org https://github.com/trunk-rs for future use. Maybe we should be proactive about the config and move it there. Thoughts?

Reply:
Anything that you're comfortable with. Should it be `dev`, `trunk-rs` and `trunk` then?
So far this is only used to safe a cache of downloaded tools. We can still change it later (until we start putting configs and other files).